Cleanup Baseapp tests

ebuchman opened the pull request:
Working on https://github.com/cosmos/cosmos-sdk/issues/1556

Will also likely address https://github.com/cosmos/cosmos-sdk/issues/1120 and replace some of https://github.com/cosmos/cosmos-sdk/pull/1172

* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [ ] Updated CHANGELOG.md
* [ ] Updated Gaia/Examples
* [x] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))
* [ ] Added appropriate labels to PR (ex. wip, ready-for-review, docs)

ebuchman commented:
Ok, this is ready aside from the changes to not run msg handlers on check tx.

It involved some fixes to the baseapp.

If you want to review the changes to baseapp_test.go, step through the commits one-by-one

ebuchman commented:
Should I rebase/squash?

ValarDragon commented on baseapp/router.go:
```diff
@@ -36,7 +36,7 @@ var isAlpha = regexp.MustCompile(`^[a-zA-Z]+$`).MatchString
 func (rtr *router) AddRoute(r string, h sdk.Handler) Router {
 	if !isAlpha(r) {
-		panic("route expressions can only contain alphanumeric characters")
+		panic("route expressions can only contain alphabet characters")
```
Good catch! However I think the regex not allowing numerics is the bug here, not the error message being mismatched.

ebuchman replied:
Let's open a new issue for this

ValarDragon commented on baseapp/baseapp_test.go:
```diff
@@ -270,12 +290,10 @@ type testTx struct {
 
 const msgType2 = "testTx"
```
Could we add a comment here saying that the point of this is that its a sample message that can fail ValidateBasic?

ebuchman commented:
@ValarDragon thanks for initial review. I rebased on develop and squashed it down to 2 commits. I think some of your comments were addressed in later commits

ebuchman commented:
It should probably just be one commit

ebuchman commented:
Ok squash down to one commit.

Let's get this merged - @cwgoes could use review on the changes to runTx for handling out-of-gas.

I'll work on #1120 in a separate PR.

ValarDragon commented on baseapp/baseapp.go:
```diff
@@ -534,53 +538,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-			}
-			return result
+		if !msgResult.IsOK() {
+			logs = append(logs, fmt.Sprintf("Msg %d failed: %s", msgIdx, msgResult.Log))
```
Does this sprintf need a `\n' at the end?

ValarDragon commented on baseapp/baseapp.go:
```diff
@@ -534,53 +538,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 
-	finalResult := sdk.Result{}
+	// accumulate results
 	var logs []string
```
Can you do `make([]string, 0, len(msgs)`, saves overhead due to expanding the array.

ValarDragon commented on baseapp/baseapp.go:
```diff
@@ -534,53 +538,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
+		ctx = app.checkState.ctx.WithTxBytes(txBytes)
+	} else {
+		ctx = app.deliverState.ctx.WithTxBytes(txBytes)
+		ctx = ctx.WithSigningValidators(app.signedValidators)
```
Maybe I'm unclear on what the purpose of simulate is, but why doesn't it get access to any of the app's signed validators? (How can it simulate how a slashing or staking tx will happen w/o this?)

ebuchman replied:
Yeh good question. The problem is that information is only supplied by BeginBlock, and when we're simulating, we don't know what the next BeginBlock will look like ...

ValarDragon replied:
Hrmm. Would it make sense to have an option where its simulated as the last tx in the last known block (for things that depend on headers / validator set), and an option for on the next block (for conceivable things that depend on the exact block height)? This discussion should probably be carried out in a separate issue though

cwgoes replied:
Could be useful, although that seems like a rare case - all slashing logic related to signing (downtime / evidence) happens in a `BeginBlock` handler, not in transactions.

cwgoes commented on baseapp/baseapp.go:
```diff
@@ -534,53 +538,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
+	// accumulate results
+	logs := make([]string, 0, len(msgs))
+	var data []byte   // NOTE: we just append them all (?!)
+	var tags sdk.Tags // also just append them all
```
https://github.com/tendermint/tendermint/pull/1803 will provide a better option here

cwgoes commented on baseapp/baseapp.go:
```diff
@@ -468,58 +475,55 @@ func (app *BaseApp) Deliver(tx sdk.Tx) (result sdk.Result) {
 		if !newCtx.IsZero() {
 			ctx = newCtx
 		}
+		gasWanted = anteResult.GasWanted
```
Presently the ante handler returns an empty `sdk.Result{}` - did we want it to set `GasWanted` (to `tx.Fee.Gas` presumably)?

ebuchman replied:
Believe so

cwgoes commented on baseapp/baseapp.go:
```diff
@@ -534,53 +538,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-	for i, msg := range msgs {
+	// accumulate results
+	logs := make([]string, 0, len(msgs))
+	var data []byte   // NOTE: we just append them all (?!)
```
I'm still not clear what exactly `data` and `logs` are for, and what the difference between them is intended to be. 

At the moment, I think the SDK uses `data` in one instance to return a proposal ID (in governance), and doesn't use logs at all. The created proposal ID could just as easily be returned as a tag (probably will be anyways, for later querying) - maybe we can simplify ABCI or what parts of ABCI we use?

ebuchman replied:
`log` is intended for non-deterministic output from the app. There is some sense that it would be used to pipe the apps actual logs back to tendermint but I don't think anyone has ever done that.

`data` is a deterministic result - we don't really use it AFAIK, so it probably could just be absorbed into tags, but it might be worth keeping for simple apps that want a definitive return value
